File: challenge_03/challenge.jl
function is_symbol(c)
    return c in ['*', '#', '+', '$']
end

function find_numbers(row)
    return match.(r"\d+", row)
end

function read_grid_from_file(filename)
    open(filename, "r") do file
        return readlines(file)
    end
end

function sum_adjacent_symbols(grid)
    total_sum = 0

    # Loop through the grid
    for i in 1:length(grid)
        row = grid[i]
        println(row)
        numbers = find_numbers(row)

        for num_match in numbers
            num_str = string(num_match.match)
            start_idx = num_match.offsets[1]
            end_idx = num_match.offsets[2]

            has_adjacent_symbol = false

            # Check left and right in the same row
            if start_idx > 1 && is_symbol(row[start_idx - 1])
                has_adjacent_symbol = true
            elseif end_idx < length(row) && is_symbol(row[end_idx + 1])
                has_adjacent_symbol = true
            end

            # Check above and below in the grid
            for k in start_idx:end_idx
                if i > 1 && is_symbol(grid[i-1][k]) || i < length(grid) && is_symbol(grid[i+1][k])
                    has_adjacent_symbol = true
                    break
                end
            end

            # Add the number if it has adjacent symbols
            if has_adjacent_symbol
                total_sum += parse(Int, num_str)
            end
        end
    end

    return total_sum
end

# Read the grid from input.txt
grid = read_grid_from_file("input.txt")

println("Sum of numbers with adjacent symbols: ", sum_adjacent_symbols(grid))

File: challenge_03/challenge.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func isSymbol(c byte) bool {
	symbols := "*#+$"
	for i := 0; i < len(symbols); i++ {
		if c == symbols[i] {
			return true
		}
	}
	return false
}

func findNumbers(row string) []string {
	re := regexp.MustCompile(`\d+`)
	return re.FindAllString(row, -1)
}

func readGridFromFile(filename string) ([]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return grid, nil
}

func main() {
	// Read the grid from input.txt
	grid, err := readGridFromFile("input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	totalSum := 0

	// Loop through the grid
	for i, row := range grid {
		numbers := findNumbers(row)
		for _, numStr := range numbers {
			startIdx := 0
			endIdx := len(numStr)

			hasAdjacentSymbol := false
			// Check left and right in the same row
			if startIdx > 0 && isSymbol(row[startIdx-1]) {
				hasAdjacentSymbol = true
			} else if endIdx < len(row) && isSymbol(row[endIdx]) {
				hasAdjacentSymbol = true
			}

			// Check above and below in the grid
			for k := startIdx; k < endIdx; k++ {
				if i > 0 && isSymbol(grid[i-1][k]) || i < len(grid)-1 && isSymbol(grid[i+1][k]) {
					hasAdjacentSymbol = true
					break
				}
			}

			// Add the number if it has adjacent symbols
			if hasAdjacentSymbol {
				num, _ := strconv.Atoi(numStr)
				totalSum += num
			}
		}
	}

	fmt.Println("Sum of numbers with adjacent symbols:", totalSum)
}
